pkg/api/tests.go

```diff
@@ -132,3 +154,129 @@ func PrintTestsJSON(release string, w http.ResponseWriter, req *http.Request, cu
 		sort(req).
 		limit(req))
 }
+
+func PrintTestsJSONFromDB(release string, w http.ResponseWriter, req *http.Request, dbc *db.DB) {
+	tests := testsAPIResult{}
+	var filter *Filter
+
+	queryFilter := req.URL.Query().Get("filter")
+	if queryFilter != "" {
+		filter = &Filter{}
+		if err := json.Unmarshal([]byte(queryFilter), filter); err != nil {
+			RespondWithJSON(http.StatusBadRequest, w, map[string]interface{}{"code": http.StatusBadRequest, "message": "Could not marshal query:" + err.Error()})
+			return
+		}
+	}
+
+	// If requesting a two day report, we make the comparison between the last
+	// period (typically 7 days) and the last two days.
+	period := req.URL.Query().Get("period")
+	if period != "current" && period != "twoDay" {
+		RespondWithJSON(http.StatusBadRequest, w, map[string]interface{}{"code": http.StatusBadRequest, "message": "Unknown period"})
+		return
+	}
+
+	testsResult, err := BuildTestsResults(dbc, release, period, filter)
+	if err != nil {
+		RespondWithJSON(http.StatusInternalServerError, w, map[string]interface{}{"code": http.StatusInternalServerError, "message": "Error building job report:" + err.Error()})
+		return
+	}
+
+	RespondWithJSON(http.StatusOK, w, testsResult.
+		sort(req).
+		limit(req))
+	RespondWithJSON(http.StatusOK, w, tests.
+		sort(req).
+		limit(req))
+}
+
+func BuildTestsResults(dbc *db.DB, release, period string, filter *Filter) (testsAPIResult, error) {
+	now := time.Now()
+
+	var testReports []apitype.Test
+	// This query takes the variants out of the picture and adds in percentages
+	q := `
+WITH results AS (
+    SELECT name,
+           sum(current_runs)       AS current_runs,
+           sum(current_successes)  AS current_successes,
+           sum(current_failures)   AS current_failures,
+           sum(current_flakes)     AS current_flakes,
+           sum(previous_runs)      AS previous_runs,
+           sum(previous_successes) AS previous_successes,
+           sum(previous_failures)  AS previous_failures,
+           sum(previous_flakes)    AS previous_flakes
+    FROM prow_test_report_7d_matview
+	WHERE release = @release
+    GROUP BY name
+)
+SELECT *,
+       current_successes * 100.0 / NULLIF(current_runs, 0) AS current_pass_percentage,
+       current_failures * 100.0 / NULLIF(current_runs, 0) AS current_failure_percentage,
+       previous_successes * 100.0 / NULLIF(previous_runs, 0) AS previous_pass_percentage,
+       previous_failures * 100.0 / NULLIF(previous_runs, 0) AS previous_failure_percentage,
+       (current_successes * 100.0 / NULLIF(current_runs, 0)) - (previous_successes * 100.0 / NULLIF(previous_runs, 0)) AS net_improvement
+FROM results;
+`
+	if period == "twoDay" {
+		q = strings.ReplaceAll(q, "prow_test_report_7d_matview", "prow_test_report_2d_matview")
```

Comment:
Can we make table/view names constants?

Reply:
Did you have specifics in mind? Were you thinking prow_job_test_report_... ?

Here's the full list today:

```
postgres=# \d+
                                                     List of relations
 Schema |                   Name                    |       Type        |  Owner   | Persistence |    Size    | Description 
--------+-------------------------------------------+-------------------+----------+-------------+------------+-------------
 public | bug_jobs                                  | table             | postgres | permanent   | 8192 bytes | 
 public | bug_tests                                 | table             | postgres | permanent   | 40 kB      | 
 public | bugs                                      | table             | postgres | permanent   | 96 kB      | 
 public | bugs_id_seq                               | sequence          | postgres | permanent   | 8192 bytes | 
 public | job_runs                                  | table             | postgres | permanent   | 8192 bytes | 
 public | job_runs_id_seq                           | sequence          | postgres | permanent   | 8192 bytes | 
 public | prow_job_run_failed_tests                 | table             | postgres | permanent   | 0 bytes    | 
 public | prow_job_run_tests                        | table             | postgres | permanent   | 756 MB     | 
 public | prow_job_run_tests_id_seq                 | sequence          | postgres | permanent   | 8192 bytes | 
 public | prow_job_runs                             | table             | postgres | permanent   | 3144 kB    | 
 public | prow_job_runs_id_seq                      | sequence          | postgres | permanent   | 8192 bytes | 
 public | prow_jobs                                 | table             | postgres | permanent   | 120 kB     | 
 public | prow_jobs_id_seq                          | sequence          | postgres | permanent   | 8192 bytes | 
 public | prow_test_analysis_by_job_14d_matview     | materialized view | postgres | permanent   | 665 MB     | 
 public | prow_test_analysis_by_variant_14d_matview | materialized view | postgres | permanent   | 201 MB     | 
 public | prow_test_report_2d_matview               | materialized view | postgres | permanent   | 37 MB      | 
 public | prow_test_report_7d_matview               | materialized view | postgres | permanent   | 37 MB      | 
 public | pull_requests                             | table             | postgres | permanent   | 8192 bytes | 
 public | pull_requests_id_seq                      | sequence          | postgres | permanent   | 8192 bytes | 
 public | release_tags                              | table             | postgres | permanent   | 8192 bytes | 
 public | release_tags_id_seq                       | sequence          | postgres | permanent   | 8192 bytes | 
 public | repositories                              | table             | postgres | permanent   | 8192 bytes | 
 public | repositories_id_seq                       | sequence          | postgres | permanent   | 8192 bytes | 
 public | suites                                    | table             | postgres | permanent   | 16 kB      | 
 public | suites_id_seq                             | sequence          | postgres | permanent   | 8192 bytes | 
 public | tests                                     | table             | postgres | permanent   | 1640 kB    | 
 public | tests_id_seq                              | sequence          | postgres | permanent   | 8192 bytes | 
(27 rows)

```

As an aside, is it worth trying to rename job_runs which is presently used for the release payload jobs?

Reply:
I totally misread, will do!